source/common/http/http1/parser_impl.cc

```diff
@@ -0,0 +1,226 @@
+          auto status = conn_impl->onUrl(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        nullptr, // on_status
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onHeaderField(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onHeaderValue(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto statusor = conn_impl->onHeadersComplete();
+          return conn_impl->setAndCheckCallbackStatusOr(std::move(statusor));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          static_cast<ParserCallbacks*>(parser->data)->bufferBody(at, length);
+          return 0;
+        },
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onMessageComplete();
+          return conn_impl->setAndCheckCallbackStatusOr(std::move(status));
+        },
+        [](llhttp_t* parser) -> int {
+          // A 0-byte chunk header is used to signal the end of the chunked body.
+          // When this function is called, http-parser holds the size of the chunk in
+          // parser->content_length. See
+          // https://github.com/nodejs/http-parser/blob/v2.9.3/http_parser.h#L336
+          const bool is_final_chunk = (parser->content_length == 0);
+          static_cast<ParserCallbacks*>(parser->data)->onChunkHeader(is_final_chunk);
+          return 0;
+        },
+        nullptr, // on_chunk_complete
+        nullptr, // on_url_complete
+        nullptr, // on_status_complete
+        nullptr, // on_header_field_complete
+        nullptr  // on_header_value_complete
+    };
+  }
+
+  RcVal execute(const char* slice, int len) {
+    llhttp_errno_t error;
+    if (slice == nullptr || len == 0) {
+      error = llhttp_finish(&parser_);
+    } else {
+      error = llhttp_execute(&parser_, slice, len);
+    }
+    size_t nread = len;
+    // Adjust number of bytes read in case of error.
+    if (error != HPE_OK) {
+      nread = llhttp_get_error_pos(&parser_) - slice;
+      // Resume after upgrade.
+      if (error == HPE_PAUSED_UPGRADE) {
+        error = HPE_OK;
+        llhttp_resume_after_upgrade(&parser_);
+      }
+    }
+    return {nread, error};
+  }
+
+  void resume() { llhttp_resume(&parser_); }
+
+  ParserStatus pause() {
+    // llhttp can only pause by returning a paused status in user callbacks.
```

Comment:
The http_parser implementation of `pause()` uses the http parser method to pause (https://github.com/nodejs/http-parser/blob/ec8b5ee63f0e51191ea43bb0c6eac7bfbff3141d/http_parser.h#L438) and an OK. That's the same as returning pause from llhttp parser. I think it's already reconciled

Reply:
I guess this is an area to cleanup once llhttp is the only implementation.

In the current world I would consider something like:
ParserStatus maybePause(ParserStatus status);

The http_parser version would detect status == Paused and call its pause method and return Success, while the llhttp version would just return the status that was passed in.

Reply:
> In the current world I would consider something like:

To me that seems strange because Paused is the only status that maybePause would ever call. 

Technically returning HPE_PAUSED from http-parser will also cause the parser to pause, but I don't trust making the change because there's very little http-parser docs and I don't know if it changes how it handles errors. (All http_parser_pause() does is cause HPE_PAUSED to be returning from execute. we could just return it directly).